Addition of Persistence Test Flow Integration with Global Test Runner

chloeminkyung opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request.

  Before submitting a pull request, please make sure the following is done:
  1. Fork [the repository](https://github.com/finos/legend-studio) and create your branch from `master`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Add a changeset to summarize your change in the changelogs.
  5. If you haven't already, complete the CLA.

  Learn more about contributing: https://github.com/finos/legend-studio/blob/master/CONTRIBUTING.md
-->

## Summary

<!--
  Explain the **motivation** for making this change. What existing problem(s) does this PR solve?
  Also, link the issue(s) to this PR: e.g. Fixes #1, Resolves #2
  If you also added documentation, link the PR from https://github.com/finos/legend
-->

## How did you test this change?

- [ ] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

<!--
  Demonstrate the code is solid. Screenshots/videos if the pull request changes the user interface.
  How exactly did you verify that your PR solves the issue you wanted to solve?
  If you leave this empty, your PR will very likely be closed.
-->

<!--
  More in-depth docs below:
  - Git workflow: https://github.com/finos/legend-studio/blob/master/docs/workflow/working-with-github.md
  - Test: https://github.com/finos/legend-studio/blob/master/docs/technical/test-strategy.md
  - Changeset: https://github.com/finos/legend-studio/blob/master/CONTRIBUTING.md#changeset
  - Dependency: https://github.com/finos/legend-studio/blob/master/docs/workflow/dependencies.md
  - UX/UI: https://github.com/finos/legend-studio/tree/master/docs/ux
-->

gayathrir11 commented:
I forgot to mention in my previous review you need to add changeset to your package. Please check the guidelines here: https://github.com/finos/legend-studio/blob/master/CONTRIBUTING.md#changeset

gayathrir11 commented on packages/legend-graph/src/graph/metamodel/pure/test/TestBatch.ts:
```diff
@@ -0,0 +1,10 @@
+import type { Hashable } from '@finos/legend-shared';
+import type { TestAssertion } from './assertion/TestAssertion.js';
+
```
please add copyright header to this

davidsunglee replied:
I have a patch for the changeset and license info. Unfortunately, I can't update Chloe's branch (she's Singapore based and not available).

Can I include in a PR immediately after merging this one? Or send one of you a patch to apply onto Chloe's PR?

akphi replied:
Is this urgent? We don't have plan to release Studio soon.

davidsunglee replied:
Trying to hit a milestone for Q3 (tomorrow). When's the next release scheduled?

davidsunglee replied:
@beekemarie FYI -- here's the final PR for the milestone

MauricioUyaguari replied:
@davidsunglee  I need to do a round of review for this which I can do tomorrow since I brought testable to studio and want to ensure I am comfortable with the changes.
@chloeminkyung could we please ensure the builds are green and add corresponding clips/screenshot for global test runner with a test batch. That is the shared component for all testable. Thanks.

davidsunglee replied:
@MauricioUyaguari Sounds good -- thanks for the update.

chloeminkyung replied:
Sorry was sick and out of office for few days. will update this by EOD

davidsunglee commented:
@MauricioUyaguari @akphi Please review when you have a moment.

@chloeminkyung confirmed that changes requested by @gayathrir11 are in.

MauricioUyaguari reviewed (changes requested):
@davidsunglee 
@chloeminkyung 
Had some comments. Most of them shouldnt take long which include:
1. lets make sure the build is green. Changesets should be up to date. Please look at our docs on what we expect on the changeset https://github.com/finos/legend-studio/blob/master/CONTRIBUTING.md#changeset as well as other PRs for examples
2. Lets make sure to add tests! Engine roundtrips are a must for these changes to ensure these flows aren't broken due to some engine changes. 
3. Let's make sure to add a short video on the global test runner addition (test batch). This helps us review css as well as your functionality. 
4. I would also consider adding to documentation registry to help your users with implementing test batches. 

The other main one is the extension mechanism for AtomicTest and TestAssertion is kinda half baked in this PR. 
I would not expect us to add to core extensions but make a PluginExtension for Testable where we can load extra things related to testable such as assertions, tests etc. 
Please take a look `StoreRelational_{Name}_Plugin_Extension` for examples. Here we load up additional things related to relational connection. `StoreRelational_PureProtocolProcessorPlugin_Extension` should be your starting point on how to implement that. 

If you have any questions on the extension comment, please set up some time with me to discuss. 

Thanks

chloeminkyung commented on packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/V1_PersistenceBuilder.ts:
```diff
@@ -621,6 +636,70 @@ export const V1_buildNotifier = (
+  parentTest: PersistenceTest | undefined,
+  context: V1_GraphBuilderContext,
+): PersistenceTestBatch[] => {
+  let persistenceTestBatches: PersistenceTestBatch[] = [];
```
Addressed

davidsunglee replied:
@MauricioUyaguari Thanks for the detailed feedback!

@chloeminkyung Hope you are feeling better, and appreciate all the updates on this one! Build failure showing a ton of quick fix lint errors. Run `yarn lint` at the project root to see them all; many can probably be auto-fixed.

chloeminkyung commented on packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/V1_PersistenceBuilder.ts:
```diff
@@ -621,6 +636,70 @@ export const V1_buildNotifier = (
+    } else {
+      throw new UnsupportedOperationError(`Can't build test assertion`, value);
+    }
+  });
```
Addressed

chloeminkyung commented on packages/legend-extension-dsl-persistence/src/graph/DSLPersistence_PureGraphPlugin.ts:
```diff
@@ -28,4 +33,18 @@ export class DSLPersistence_PureGraphPlugin extends PureGraphPlugin {
+  override getExtraTestablesCollectors(): TestablesCollector[] {
+    return [
+      (graph: PureModel): Testable[] => {
+        const testables: Testable[] = [];
```
Addressed

chloeminkyung commented on packages/legend-graph/src/index.ts:
```diff
@@ -180,6 +180,7 @@ export * from './graph/helpers/creator/StoreRelational_ModelCreatorHelper.js';
 
 export * from './graphManager/helpers/DSLData_GraphManagerHelper.js';
 export * from './graphManager/helpers/ValueSpecificationGraphManagerHelper.js';
+export * from './graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestBuilderHelper.js';
```
Addressed!

chloeminkyung commented on packages/legend-application-studio/src/components/editor/side-bar/testable/GlobalTestRunner.tsx:
```diff
@@ -340,6 +341,13 @@ const TestableTreeNodeContainer: React.FC<
           </div>
         )}
+        {node instanceof TestBatchTreeNodeData && (
+          <div className="global-test-runner__item__link__content">
```
Added

chloeminkyung commented:
@MauricioUyaguari  
@davidsunglee 

Thanks for the comments Mauricio!

Most of the PR-related comments are addressed:
- [x] Changesets updated
- [x] Rebased
- [x] Description added
- [x] tests in manual-test and roundtrips added
- [x] short video on the global test runner addition (test batch) added

However, few questions around followings:
1. Is there a fixed place for a documentation registry for me to add details of implementing test batches?
2. Not sure what do you mean by half baked, is the extension mechanism related around refactoring of codes? AtomicTest builders and transformers are already in the extension using getExtra.. functions (correct me if I'm wrong @gayathrir11) - if it's insufficient, I probably need more details around the best practice :)


Thank you!

chloeminkyung commented:
> @MauricioUyaguari @davidsunglee
> 
> Thanks for the comments Mauricio!
> 
> Most of the PR-related comments are addressed:
> 
> * [x]  Changesets updated
> * [x]  Rebased
> * [x]  Description added
> * [ ]  tests in manual-test and roundtrips added
> * [x]  short video on the global test runner addition (test batch) added
> 
> However, few questions around followings:
> 
> 1. Is there a fixed place for a documentation registry for me to add details of implementing test batches?
> 2. Not sure what do you mean by half baked, is the extension mechanism related around refactoring of codes? AtomicTest builders and transformers are already in the extension using getExtra.. functions (correct me if I'm wrong @gayathrir11) - if it's insufficient, I probably need more details around the best practice :)
> 
> Thank you!

Fixing roundTrip test a bit

MauricioUyaguari commented:
> @MauricioUyaguari @davidsunglee
> 
> Thanks for the comments Mauricio!
> 
> Most of the PR-related comments are addressed:
> 
> * [x]  Changesets updated
> * [x]  Rebased
> * [x]  Description added
> * [ ]  tests in manual-test and roundtrips added (fixing some issue atm)
> * [x]  short video on the global test runner addition (test batch) added
> 
> However, few questions around followings:
> 
> 1. Is there a fixed place for a documentation registry for me to add details of implementing test batches?
> 2. Not sure what do you mean by half baked, is the extension mechanism related around refactoring of codes? AtomicTest builders and transformers are already in the extension using getExtra.. functions (correct me if I'm wrong @gayathrir11) - if it's insufficient, I probably need more details around the best practice :)
> 
> Thank you!

1. Thanks. I think we need another rebase :)
2. The extension mechanism for Testable's test should be done more like the relational connection extension that I mentioned. For now I think I can merge this PR if you rebase. We will fix the extension mechanism for testable test as it should of been done during the initial Testable PR.

chloeminkyung commented:
> > @MauricioUyaguari @davidsunglee
> > Thanks for the comments Mauricio!
> > Most of the PR-related comments are addressed:
> > 
> > * [x]   Changesets updated
> > * [x]   Rebased
> > * [x]   Description added
> > * [ ]   tests in manual-test and roundtrips added (fixing some issue atm)
> > * [x]   short video on the global test runner addition (test batch) added
> > 
> > However, few questions around followings:
> > 
> > 1. Is there a fixed place for a documentation registry for me to add details of implementing test batches?
> > 2. Not sure what do you mean by half baked, is the extension mechanism related around refactoring of codes? AtomicTest builders and transformers are already in the extension using getExtra.. functions (correct me if I'm wrong @gayathrir11) - if it's insufficient, I probably need more details around the best practice :)
> > 
> > Thank you!
> 
> 1. Thanks. I think we need another rebase :)
> 2. The extension mechanism for Testable's test should be done more like the relational connection extension that I mentioned. For now, I think I can merge this PR if you rebase. We will fix the extension mechanism for testable test as it should of been done during the initial Testable PR. To do this though I do need the manual tests added to ensure stability. Can you confirm those run and are added. I did see them in the PR but double checking.

Rebased and fixed the tests so should be ready by now. Let us know if anything looks off.

MauricioUyaguari commented:
@chloeminkyung @davidsunglee 
1. Changeset is incorrect. Any reason why the diffs are showing previous changesets as additions? 
2. Also I dont see changeset messages specific to your change. 
I would expect to see one tied to your persistance dsl such as: 

```
'@finos/legend-extension-dsl-persistence': minor
---
Add persistance test graph support. 
```

and another tied to adding TestBatch to the application such as:


```
'@finos/legend-application-studio': minor
'@finos/legend-graph': minor
---
Add `TestBatch` to Testable model and incorporate into GlobalTestRunner. 
```

and a third one with the remaining packages you changed with an empty message.

MauricioUyaguari commented on packages/legend-manual-tests/src/__tests__/roundtrip-grammar/cases/DSL_Persistence-basic.pure:
```diff
@@ -172,6 +172,10 @@ Persistence org::legend::ZooPersistence
       }
     ];
   }
+  tests:
```
You may want to fix this in `engine` composer
You don't want to render the tests part of the grammar if there is no tests.

chloeminkyung replied:
Isn't TestSuite the same? tried to have some consistency over there.
